### PR TITLE
[mqtt] Restart upstream bridge when stopped

### PR DESCRIPTION
**Pull request by dmolokanov**
restart the upstream bridge when it is stopped unexpectedly.

### CODE REVIEW

File: mqtt/mqttd/src/broker/bootstrap/edgehub.rs [hunk 1]
```diff
@@ -171,6 +171,8 @@ pub async fn start_sidecars(
     let system_address = listener_settings.system().addr().to_string();
 
     let bridge_controller = BridgeController::new();
+    // TODO implement shutdown for bridge
+    let _bridge_controller_handle = bridge_controller.handle();
```
Comment (dmolokanov) [inline mqtt/mqttd/src/broker/bootstrap/edgehub.rs:175]: I'll address this comment in the next PR

File: mqtt/mqtt-bridge/src/config_update.rs [hunk 2]
```diff
@@ -129,8 +129,26 @@ pub struct BridgeUpdate {
...
+            .chain(forwards.into_iter().map(Direction::Out))
+            .collect();
+
+        Self {
+            endpoint: name.into(),
+            subscriptions,
+        }
+    }
+
+    // TODO update should have name
```
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/config_update.rs:145]: do you have the full spec of the twin schema?

File: mqtt/mqtt-bridge/src/controller/mod.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        let mut no_bridges = bridges.is_terminated();
+
+        loop {
+            let wait_bridge_or_pending = if no_bridges {
+                // if no active bridges available, wait only for a new messages arrival
+                Either::Left(future::pending())
+            } else {
+                // otherwise try to await both a new message arrival or any bridge exit
+                Either::Right(bridges.next())
+            };
```
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/controller/mod.rs:76]: actually no, we should check that **each** loop iteration. 
the controller is reactive by nature: it reacts on the twin update which can be the one to remove all bridges and later to bridge some back

File: mqtt/mqtt-bridge/src/controller/mod.rs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+                    process_update(update, &mut bridges).await
+                }
+                Either::Left((BridgeControllerMessage::Shutdown, _)) => {
+                    info!("bridge controller shutdown requested");
+                    bridges.shutdown_all().await;
+                    break;
+                }
+                Either::Right((Some((name, bridge)), _)) => {
+                    match bridge {
+                        Ok(Ok(_)) => debug!("bridge {} exited", name),
```
[the remark this answers is not in the document]
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/controller/mod.rs:89]: I guess it is true for the upstream bridge, and not for any bridge.

File: mqtt/mqtt-bridge/src/controller/mod.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+#[derive(Debug)]
+pub enum BridgeControllerMessage {
+    BridgeControllerUpdate(BridgeControllerUpdate),
+    Shutdown,
+}
+
+/// Error for `BridgeController`.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("An error occurred sending a message to the controller.")]
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/controller/mod.rs:181]: `{0}` for all errors

File: mqtt/mqtt-bridge/src/controller/mod.rs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+                    match bridge {
+                        Ok(Ok(_)) => debug!("bridge {} exited", name),
+                        Ok(Err(e)) => warn!(error = %e, "bridge {} exited with error", name),
+                        Err(e) => warn!(error = %e, "bridge {} panicked ", name),
+                    }
+
+                    // always restart upstream bridge
+                    if name == UPSTREAM {
+                        info!("restarting bridge {}...", name);
+                        if let Some(upstream_settings) = settings.upstream() {
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/controller/mod.rs:98]: nit: looks very similar to the code above. Extract func?

File: mqtt/mqtt-bridge/src/controller/bridges.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        debug!("sending shutdown request to all bridges...");
+
+        // sending shutdown signal to each bridge
+        let shutdowns = self
+            .bridge_handles
+            .drain()
+            .map(|(_, handle)| handle.shutdown());
+        future::join_all(shutdowns).await;
+
+        debug!("waiting for all bridges to exit...");
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/controller/bridges.rs:78]: this message makes sense before `join_all`, right?

File: mqtt/mqtt-bridge/src/controller/bridges.rs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    pub(crate) async fn send_update(&mut self, update: BridgeUpdate) {
+        if let Some(config) = self.config_updaters.get_mut(update.name()) {
+            if let Err(e) = config.send_update(update).await {
+                error!("error sending bridge update {:?}", e);
+            }
+        }
+    }
+
+    pub(crate) async fn shutdown_all(&mut self) {
+        debug!("sending shutdown request to all bridges...");
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/src/controller/bridges.rs:69]: nit: maybe `info!` level is fine here?

**Review by vadim-kovalyov — APPROVED** [review]
LGTM, just few nits

File: mqtt/mqtt-bridge/src/controller/bridges.rs (hunk 7 above)
Comment (dmolokanov) [reply mqtt/mqtt-bridge/src/controller/bridges.rs:78]: actually now. the log message before sending shutdown signal on the line above actully :)

**Pull request merged**